File: product-service/src/main/java/com/iuh/entity/Book.java
package com.iuh.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.iuh.enums.BookStatus;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;

import java.util.HashSet;
import java.util.Set;

@Setter
@Getter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
@Entity
@Table(name = "books")
public class Book extends AbstractEntity {

    @Column(nullable = false)
    String title;

    @Column(nullable = false, unique = true)
    String slug;

    @Column(nullable = false, columnDefinition = "TEXT")
    String thumbnail;

    @Column(columnDefinition = "TEXT")
    String description;

    String author;

    String size;

    Integer pages;

    Integer weight;

    Integer publishYear;

    Double importPrice;

    Double price;

    Integer stock;

    Integer sold;

    Integer reviewCount;

    Double reviewStar;

    Boolean isNew;

    Boolean isFeatured;

    @Enumerated(EnumType.STRING)
    @Builder.Default
    BookStatus status = BookStatus.DRAFT;

    @ManyToOne()
    @JoinColumn(name = "category_id")
    Category category;

    @ManyToOne()
    @JoinColumn(name = "publisher_id")
    Publisher publisher;

    @ManyToOne()
    @JoinColumn(name = "discount_id")
    Discount discount;

    @JsonIgnore
    @OneToMany(mappedBy = "book", cascade = CascadeType.ALL, orphanRemoval = true)
    Set<BookImage> bookImages;

    public void addBookImage(BookImage bookImage) {
        if (bookImages == null) {
            bookImages = new HashSet<>();
        }
        bookImages.add(bookImage);
        bookImage.setBook(this);
    }

    public Double getDiscountPrice() {
        if (discount == null || !discount.isValid()) {
            return price;
        }
        return price - (price * discount.getPercent() / 100);
    }
}

File: web-app/types/book.ts
import { Book } from 'lucide-react'

export type BookImage = {
  id: string
  url: string
}

export type Category = {
  id: string
  name: string
  slug: string
  description?: string
  createdAt: string
  updatedAt: string
}

export type Publisher = {
  id: string
  name: string
  slug: string
  description?: string
  createdAt: string
  updatedAt: string
}

export type Book = {
  id: string
  title: string
  slug: string
  thumbnail: string
  description: string
  author: string
  size: string
  pages: number
  weight: number
  publishYear: number
  importPrice: number
  price: number
  discountPrice: number
  stock: number
  sold: number
  reviewCount: number
  reviewStar: number
  discount?: Discount
  isNew: boolean
  isFeatured: boolean
  status: BookStatus
  category: Category
  publisher: Publisher
  bookImages: BookImage[]
  createdAt: string
  updatedAt: string
}

export enum BookStatus {
  ACTIVE = 'ACTIVE',
  DRAFT = 'DRAFT',
  DISABLED = 'DISABLED',
}

export type BookResponse = {
  code: number
  message: string
  data: Book
}

export type PageBooksResponse = {
  code: number
  message: string
  data: {
    pageNo: number
    pageSize: number
    totalPages: number
    totalElements: number
    items: Book[]
  }
}

export type PageBooks = {
  pageNo: number
  pageSize: number
  totalPages: number
  totalElements: number
  items: Book[]
}


export interface BookSearchParams {
  pageNo?: string
  pageSize?: string
  sortBy?: string
  books?: string
}

export type Discount = {
  id: string
  createdAt: string
  updatedAt: string
  name: string
  code: string
  percent: number
  startDate: string
  endDate: string
}